source/agora/utils/Test.d

```diff
@@ -624,62 +627,274 @@ unittest
+
+        Signs the inputs.
+
+        Params:
+            type = type of `Transaction`
+            unknowns = unknown `KeyPair`s.
+                finds in `unknowns` first. If not, finds it in 'WK'.
+
+        Returns:
+            returns the final built `Transaction`.
+
+    ***************************************************************************/
+
+    public Transaction sign (TxType type = TxType.Payment,
+        scope const(KeyPair)[] unknowns = null) @safe
+    {
+        assert(this.inputs.length, "Cannot sign input-less transaction");
+        assert(this.data.outputs.length, "Onput-less transactions are not valid");
+
+        this.data.type = type;
+
+        // Finalize the transaction by adding inputs
+        foreach (ref in_; this.inputs)
+            this.data.inputs ~= Input(in_.hash(), in_.index);
+
+        // Get the hash to sign
+        const txHash = this.data.hashFull();
+        // Sign all inputs using WK keys
+        foreach (idx, ref in_; this.inputs)
+        {
+            auto ownerKP = getOwnerKeyPair(in_.output.address, unknowns);
+            assert(ownerKP !is KeyPair.init,
+                    "Address not found in Well-Known keypairs: "
+                    ~ in_.output.address.toString());
+            auto signature = () @trusted { return ownerKP.secret.sign(txHash[]); }();
+            this.data.inputs[idx].signature = signature;
+        }
+
+        // Return the result and reset this
+        this.inputs = null;
+        scope (success) this.data = Transaction.init;
+        return this.data;
+    }
+
+    /***************************************************************************
 
-    auto remainder = amount.div(toward.length);
-    Transaction result = Transaction(type);
-    foreach (addr; toward)
-        result.outputs ~= Output(amount, addr);
+        Finds a well-known `KeyPair` corresponding to the `PublicKey`.
 
-    // Add the remainder to the first output.
-    result.outputs[0].value.mustAdd(remainder);
-    const inputHash = input.hashFull();
+        Params:
+            address = finding `PublicKey`
+            unknowns =  unknown `KeyPair`s
 
-    // Add support for Transactions with multiple recipients
-    foreach (idx, const ref _; input.outputs)
-        result.inputs ~= Input(inputHash, cast(uint) idx);
+        Returns:
+            returns KeyPair.init if `address` is not found.
 
-    scope sign = (KeyPair kp, Hash h) @trusted { return kp.secret.sign(h[]); };
+    ***************************************************************************/
 
-    const resultHash = result.hashFull();
-    foreach (idx, ref in_; result.inputs)
+    private KeyPair getOwnerKeyPair (scope const PublicKey address,
+        scope const(KeyPair)[] unknowns = null) @safe
     {
-        auto rng = from.find!(a => a.address == input.outputs[idx].address);
-        assert(rng.length);
-        const owner = rng[0];
-        in_.signature = sign(owner, resultHash);
+        version(unittest)
```

Comment:
This whole module is made to be used from tests so having this special case is a bit surprising

Reply:
I see why it was introduced but I think there are better ways. Let's discuss a bit later.